cpp/src/fil/fil.cu

```diff
@@ -71,43 +71,63 @@ __global__ void transform_k(float* preds, size_t n, output_t output,
 }
 
 struct forest {
-  void init_max_shm() {
+  void enumerate_n_items() {
+    for (ssp_.n_items = algo_ == algo_t::BATCH_TREE_REORG ? 4 : 1;
+         ssp_.n_items != 0 &&
+         ssp_.max_shm < (ssp_.shm_sz = get_smem_footprint(ssp_, leaf_algo_));
+         ssp_.n_items--)
+      ;
+  }
+
+  void init_n_items(int device) {
     int max_shm_std = 48 * 1024;  // 48 KiB
-    int device = 0;
-    // TODO(canonizer): use raft::handle_t for this
-    CUDA_CHECK(cudaGetDevice(&device));
     CUDA_CHECK(cudaDeviceGetAttribute(
-      &max_shm_, cudaDevAttrMaxSharedMemoryPerBlockOptin, device));
+      &ssp_.max_shm, cudaDevAttrMaxSharedMemoryPerBlockOptin, device));
     // TODO(canonizer): use >48KiB shared memory if available
-    max_shm_ = std::min(max_shm_, max_shm_std);
+    ssp_.max_shm = std::min(ssp_.max_shm, max_shm_std);
+
+    /** searching for the most items per block while respecting the shared
+    * memory limits creates a full linear programming problem.
+    * solving it in a single equation looks less tractable than this */
```

Comment:
```suggestion
    // solving it in a single equation looks less tractable than this
```

This is not a documentation comment, so please avoid using the documentation comment style for it.